include/cppflow/span.h

```diff
@@ -0,0 +1,588 @@
+};
+
+#ifdef TCB_SPAN_HAVE_DEDUCTION_GUIDES
+
+/* Deduction Guides */
+template <class T, size_t N>
+    span(T (&)[N])->span<T, N>;
+
+template <class T, size_t N>
+    span(std::array<T, N>&)->span<T, N>;
+
+template <class T, size_t N>
+    span(const std::array<T, N>&)->span<const T, N>;
+
+template <class Container>
+    span(Container&)->span<typename Container::value_type>;
+
+template <class Container>
+    span(const Container&)->span<const typename Container::value_type>;
+
+#endif // TCB_HAVE_DEDUCTION_GUIDES
+
+template <typename ElementType, std::size_t Extent>
+    constexpr span<ElementType, Extent>
+make_span(span<ElementType, Extent> s) noexcept {
+    return s;
+}
+
+template <typename T, std::size_t N>
+constexpr span<T, N> make_span(T (&arr)[N]) noexcept {
+    return {arr};
+}
+
+template <typename T, std::size_t N>
+TCB_SPAN_ARRAY_CONSTEXPR span<T, N> make_span(std::array<T, N>& arr) noexcept {
+    return {arr};
+}
+
+template <typename T, std::size_t N>
+    TCB_SPAN_ARRAY_CONSTEXPR span<const T, N>
+make_span(const std::array<T, N>& arr) noexcept {
+    return {arr};
+}
+
+template <typename Container>
+constexpr span<typename Container::value_type> make_span(Container& cont) {
+    return {cont};
+}
+
+template <typename Container>
+    constexpr span<const typename Container::value_type>
+make_span(const Container& cont) {
+    return {cont};
+}
+
+template <typename ElementType, std::size_t Extent>
+    span<const byte, ((Extent == dynamic_extent) ? dynamic_extent
+            : sizeof(ElementType) * Extent)>
+as_bytes(span<ElementType, Extent> s) noexcept {
+    return {reinterpret_cast<const byte*>(s.data()), s.size_bytes()};
+}
+
+template <class ElementType, size_t Extent,
+          typename std::enable_if<!std::is_const<ElementType>::value, int>::type = 0>
+              span<byte, ((Extent == dynamic_extent) ? dynamic_extent
+                      : sizeof(ElementType) * Extent)>
+as_writable_bytes(span<ElementType, Extent> s) noexcept {
+  return {reinterpret_cast<byte*>(s.data()), s.size_bytes()};
+}
+
+template <std::size_t N, typename E, std::size_t S>
+constexpr auto get(span<E, S> s) -> decltype(s[N]) {
+    return s[N];
+}
+
+} // namespace TCB_SPAN_NAMESPACE_NAME
+
+namespace std {
+
+template <typename ElementType, size_t Extent>
+class tuple_size<TCB_SPAN_NAMESPACE_NAME::span<ElementType, Extent>>
+: public integral_constant<size_t, Extent> {};
+
+template <typename ElementType>
+class tuple_size<TCB_SPAN_NAMESPACE_NAME::span<
+ElementType, TCB_SPAN_NAMESPACE_NAME::dynamic_extent>>; // not defined
+
+template <size_t I, typename ElementType, size_t Extent>
+class tuple_element<I, TCB_SPAN_NAMESPACE_NAME::span<ElementType, Extent>> {
+public:
+    static_assert(Extent != TCB_SPAN_NAMESPACE_NAME::dynamic_extent &&
+            I < Extent,
+            "");
+    using type = ElementType;
+};
+
+
+} // end namespace std
+
+#ifdef __has_cpp_attribute
+#  if __has_cpp_attribute(__cpp_lib_span)
```

Comment:
I don't think this will work. According to <https://en.cppreference.com/w/cpp/feature_test>, "Library features" section:
`The following macros are defined if the header <version> or any of the corresponding headers in the table below is included.`
which means `__cpp_lib_span` is only defined if `<span>` or `<version>` is included. We need to test `__has_include(<span>)` and include it before testing `__cpp_lib_span`. There is example code in the bottom of the previous link.